### ISSUE TITLE
manage users / clients connected outside server file

### ISSUE THREAD

**Issue by gfpinpact** [walkor/phpsocket.io#147]
Hi, walkor, great Job.

any idea how access to clients connected outside server file?

i used laravel 5.6, and my server file is 

```PHP
use Workerman\Worker;
use Workerman\WebServer;
use Workerman\Autoloader;
use PHPSocketIO\SocketIO;
use App\Socket;

// composer autoload
autoLoad();

$SSL = env('SOCKET_SSL', false);
$PORT = env('SOCKET_PORT', 9000);

if ($SSL) {
    $context = array(
        'ssl' => array(
            'local_cert' => '/etc/letsencrypt/live/XX/cert.pem',
            'local_pk' => '/etc/letsencrypt/live/XX/privkey.pem',
            'verify_peer' => false,
        )
    );
    $io = new SocketIO($PORT, $context);
} else {
    $io = new SocketIO($PORT);
}

$socket_layer_controller = new Socket\SocketLayerController($io);

$io->on('workerStart', function($socket) {
    // on start
});

$io->on('connection', function($socket) use ($socket_layer_controller) {
    $functions = \get_class_methods($socket_layer_controller);

    $socket_layer_controller->newClientConnected($socket);

    foreach ($functions as $function) {
        if (substr($function, 0, 2) == 'io') {
            $function_str = getFunctionString($function);
            $socket->on($function_str, function ($data) use ($socket, $socket_layer_controller, $function, $function_str) {
                call_user_func_array([$socket_layer_controller, $function], [$data, $socket, $function_str]);
            });
        }
    }
}

```

And my controller layer (for write events outside file server) is 

```PHP
namespace App\Socket;

class SocketLayerController extends SocketBase {

    const EVENT_LOGIN_OK = 'login-success';
    const EVENT_LOGIN_ERROR = 'login-error';
    const EVENT_LOGOUT_OK = 'logout-ok';
    const EVENT_CLIENT_CONNECTED = 'join-client';

    public function __construct(\PHPSocketIO\SocketIO $io) {
        parent::__construct($io);
    }

    public function newClientConnected($socket) {
        try {
            $new_cliente = new SocketClient($socket);
           \Socket::addClient($new_cliente);
            
        } catch (\Exception $ex) {
            print_r($ex->getMessage());
        }
    }

    public function ioLogin($data, $client, $event_string) {
        try {
//            if ($data['key'] == 'XXXX') {
                $this->emitTo($client->id, self::EVENT_LOGIN_OK, $client->id);
                $this->emitDataAll(self::EVENT_CLIENT_CONNECTED, $client->id);
//            } else {
//                $this->emitTo($client->id, self::EVENT_LOGIN_ERROR, null);
//                $client->disconnect(true);
//            }
        } catch (\Exception $ex) {
            print_r($ex->getMessage());
        }
    }

    public function ioLogout($data, $client, $event_string) {
        try {
            $this->emitTo($client->id, self::EVENT_LOGOUT_OK, null);
            $client->disconnect(true);
        } catch (\Exception $ex) {
            print_r($ex->getMessage());
        }
    }

    public function ioMessage($data, $client, $event_string) {
        try {
            $this->emitDataAll($event_string, $this->clients);
        } catch (\Exception $ex) {
            print_r($ex->getMessage());
        }
    }
```
my problem is when I need to access the connected clients from the laravel controllers.
something like this :

```PHP
\Socket::getClients()
```
it's understandable?

PD: sorry for my english.

**Comment by LuckyCyborg** [comment]
You can access the connected clients from a SocketIO server withing a Laravel routed Controller?

I guess, not. **Never**. Because the SocketIO server runs, well... over a particular WebSocket based protocol, which is not a HTTP request which you can dispatch to controllers from Laravel or whatever.

Expect from PHPSocketIO just what gives SocketIO, only that the first one is written in PHP. 

Eventually, you can write your own (PHP)SocketIO server, then to talk somehow with it, via the SocketIO client javascript. 

Imagine something similar with the Echo, supposing that you write a Echo Server on top of PHPSocketIO.

Or you can adopt a Pusher like style. :smirk: 

If you are curious how can look a Pusher-like server made in top of PHPSocketIO and supporting multiple sites client, which probably you can use also with Laravel, if you write a proper client side, you can look at the **Quasar Push Server**: https://github.com/nova-framework/quasar written for **Nova Framework**.

To note that that's a work in progress and the javascript client part is there: https://github.com/nova-framework/framework/blob/4.0/assets/js/quasar.js

**Comment by gfpinpact** [comment]
Dear,
First, thanks for such a quick response.

I found a simple solution, which I hope will be useful in the long term. Memcached.

With memcached I keep the IDs of the connected clients and then from the laravel drivers I send messages to all the channels / users or to one in specific (for that I needed to save the id's).

file server:
  - save ids clients on memcached.

controllers laravel:
  - get clients id's from memcached
  - emmit message to specify id or all clients.

for now I do not need more.

I will investigate the information that you have given me, some time ago work with quasar, I hope now this more advanced.
